keps/sig-instrumentation/647-apiserver-tracing/README.md

```diff
@@ -0,0 +1,226 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This Kubernetes Enhancement Proposal (KEP) proposes enhancing the API Server to allow tracing requests.  For this, it proposes using OpenTelemetry libraries, and exports in the OpenTelemetry format.
+
+## Motivation
+
+Along with metrics and logs, traces are a useful form of telemetry to aid with debugging incoming requests.  The API Server currently uses a poor-man's form of tracing (see [github.com/kubernetes/utils/trace](https://github.com/kubernetes/utils/tree/master/trace)), but we can make use of distributed tracing to improve the ease of use and enable easier analysis of trace data.  Trace data is structured, providing the detail necessary to debug requests, and context propagation allows plugins, such as admission webhooks, to add to API Server requests.
+
+### Definitions
+
+**Span**: The smallest unit of a trace.  It has a start and end time, and is attached to a single trace.
+**Trace**: A collection of Spans which represents a single process.
+**Trace Context**: A reference to a Trace that is designed to be propagated across component boundaries.  Sometimes referred to as the "Span Context".  It is can be thought of as a pointer to a parent span that child spans can be attached to.
+
+### Goals
+
+* The API Server generates and exports spans for incoming and outgoing requests.
+* The API Server propagates context from incoming requests to outgoing requests.
+
+### Non-Goals
+
+* Tracing in kubernetes controllers
+* Replace existing logging, metrics, or the events API
+* Trace operations from all Kubernetes resource types in a generic manner (i.e. without manual instrumentation)
+* Change metrics or logging (e.g. to support trace-metric correlation)
+* Access control to tracing backends
+* Add tracing to components outside kubernetes (e.g. etcd client library).
+
+## Proposal
+
+### Tracing API Requests
+
+We will wrap the API Server's http server and http clients with [otelhttp](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/master/instrumentation/net/http/otelhttp) to get spans for incoming and outgoing http requests.  This generates spans for all sampled incoming requests and propagates context with all client requests.  For incoming requests, this would go below [WithRequestInfo](https://github.com/kubernetes/kubernetes/blob/9eb097c4b07ea59c674a69e19c1519f0d10f2fa8/staging/src/k8s.io/apiserver/pkg/server/config.go#L676) in the filter stack, as it must be after authentication and authorization, before the panic filter, and is closest in function to the WithRequestInfo filter.
+
+Note that some clients of the API Server, such as webhooks, may make reentrant calls to the API Server.  To gain the full benefit of tracing, such clients should propagate context with requests back to the API Server.
```

Comment:
Please expand this, how do clients do that?

Reply:
I have an extremely strong interest in apiserver knowing when an incoming API call is re-entrant, and ideally, what its parent call is. Additionally, I'd like this to work in a multiple apiserver setup. So:
* Can the tracing header be reverse-engineered to figure this out?
* Can I pack metadata into the tracing header to make this possible?

Reply:
I gave an example of how a webhook would do that.  The tracing header will tell you if it is a direct child of the API Server trace, but you can't tell if it is an indirect child (e.g. if the admission controller webhook adds a span in-between).  You can pack metadata into the tracing header (well, it is a different header, but still), using "Baggage": https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/baggage/api.md#baggage-api.

Reply:
Thanks, love the name "Baggage"